### ISSUE TITLE
Loss got higher and became NaN starting from 600 steps

### ISSUE THREAD

**Issue by NyanSwanAung** [TannerGilbert/Tensorflow-Object-Detection-API-train-custom-Mask-R-CNN-model#13]
**Describe the bug**
my loss got inclining after 500 steps and became nan by the time it reached 600 steps. How did you manage to train 3000 steps with great accuracy? I saw your protocol buffer file in this repo but I just wanna know. 

**Expected behavior**
a good accuracy and loss

** Current Results trained with 300 steps, loss = 0.45572507 before inclining loss ** 
<img width="399" alt="Screen Shot 2021-06-06 at 15 08 18" src="https://user-images.githubusercontent.com/48433214/120918257-88c8dc80-c6d9-11eb-8f05-8888a549fdf6.png">
<img width="399" alt="Screen Shot 2021-06-06 at 15 08 34" src="https://user-images.githubusercontent.com/48433214/120918262-8cf4fa00-c6d9-11eb-89d5-64a2b92d61f2.png">
<img width="399" alt="Screen Shot 2021-06-06 at 15 08 44" src="https://user-images.githubusercontent.com/48433214/120918264-91b9ae00-c6d9-11eb-964d-30dd928b6198.png">

**Config File**
``` 
Step 100 per-step time 18.226s
{'Loss/BoxClassifierLoss/classification_loss': 0.014474193,
 'Loss/BoxClassifierLoss/localization_loss': 0.20782878,
 'Loss/BoxClassifierLoss/mask_loss': 2.7738497,
 'Loss/RPNLoss/localization_loss': 0.2956658,
 'Loss/RPNLoss/objectness_loss': 0.6722792,
 'Loss/regularization_loss': 0.0,
 'Loss/total_loss': 3.9640975,
 'learning_rate': 0.00016000001}

Step 200 per-step time 17.014s
{'Loss/BoxClassifierLoss/classification_loss': 0.028602576,
 'Loss/BoxClassifierLoss/localization_loss': 0.0,
 'Loss/BoxClassifierLoss/mask_loss': 0.0,
 'Loss/RPNLoss/localization_loss': 0.011901135,
 'Loss/RPNLoss/objectness_loss': 0.61675644,
 'Loss/regularization_loss': 0.0,
 'Loss/total_loss': 0.6572602,
 'learning_rate': 0.00032000002}

Step 300 per-step time 16.788s
{'Loss/BoxClassifierLoss/classification_loss': 0.016163696,
 'Loss/BoxClassifierLoss/localization_loss': 0.0,
 'Loss/BoxClassifierLoss/mask_loss': 0.0,
 'Loss/RPNLoss/localization_loss': 0.030279242,
 'Loss/RPNLoss/objectness_loss': 0.40928215,
 'Loss/regularization_loss': 0.0,
 'Loss/total_loss': 0.45572507,
 'learning_rate': 0.00048000005}

Step 400 per-step time 16.973s
{'Loss/BoxClassifierLoss/classification_loss': 0.0,
 'Loss/BoxClassifierLoss/localization_loss': 17.068367,
 'Loss/BoxClassifierLoss/mask_loss': 2.7716024,
 'Loss/RPNLoss/localization_loss': 0.016921213,
 'Loss/RPNLoss/objectness_loss': 0.043970272,
 'Loss/regularization_loss': 0.0,
 'Loss/total_loss': 19.900862,
 'learning_rate': 0.00064000004}

Step 500 per-step time 16.409s
'Loss/BoxClassifierLoss/classification_loss': 44692950000.0,
 'Loss/BoxClassifierLoss/localization_loss': 29256022000.0,
 'Loss/BoxClassifierLoss/mask_loss': 6294802.5,
 'Loss/RPNLoss/localization_loss': 0.04560062,
 'Loss/RPNLoss/objectness_loss': 0.04735045,
 'Loss/regularization_loss': 0.0,
 'Loss/total_loss': 73955270000.0,
 'learning_rate': 0.00080000004}

Step 600 per-step time 15.125s
{'Loss/BoxClassifierLoss/classification_loss': nan,
 'Loss/BoxClassifierLoss/localization_loss': nan,
 'Loss/BoxClassifierLoss/mask_loss': nan,
 'Loss/RPNLoss/localization_loss': nan,
 'Loss/RPNLoss/objectness_loss': nan,
 'Loss/regularization_loss': 0.0,
 'Loss/total_loss': nan,
 'learning_rate': 0.0009600001}

Step 700 per-step time 14.040s
{'Loss/BoxClassifierLoss/classification_loss': nan,
 'Loss/BoxClassifierLoss/localization_loss': nan,
 'Loss/BoxClassifierLoss/mask_loss': nan,
 'Loss/RPNLoss/localization_loss': nan,
 'Loss/RPNLoss/objectness_loss': nan,
 'Loss/regularization_loss': 0.0,
 'Loss/total_loss': nan,
 'learning_rate': 0.0011200001}
```
**Desktop (please complete the following information):**
 - OS: [MacOS Big Sur]
 - Training on jupyter notebook
 - GPU: [Intel Iris Plus Graphics 655 1536MB]
 - CPU : [Intel(R) Core(TM) i5-8259U CPU @ 2.30GHz]

**Config Details:**  
``` 
model {
  faster_rcnn {
    number_of_stages: 3
    num_classes: 1
    image_resizer {
      fixed_shape_resizer {
        height: 1024
        width: 1024
      }
    }
    feature_extractor {
      type: "faster_rcnn_inception_resnet_v2_keras"
    }
    first_stage_anchor_generator {
      grid_anchor_generator {
        height_stride: 16
        width_stride: 16
        scales: 0.25
        scales: 0.5
        scales: 1.0
        scales: 2.0
        aspect_ratios: 0.5
        aspect_ratios: 1.0
        aspect_ratios: 2.0
      }
    }
    first_stage_box_predictor_conv_hyperparams {
      op: CONV
      regularizer {
        l2_regularizer {
          weight: 0.0
        }
      }
      initializer {
        truncated_normal_initializer {
          stddev: 0.01
        }
      }
    }
    first_stage_nms_score_threshold: 0.0
    first_stage_nms_iou_threshold: 0.7
    first_stage_max_proposals: 300
    first_stage_localization_loss_weight: 2.0
    first_stage_objectness_loss_weight: 1.0
    initial_crop_size: 17
    maxpool_kernel_size: 1
    maxpool_stride: 1
    second_stage_box_predictor {
      mask_rcnn_box_predictor {
        fc_hyperparams {
          op: FC
          regularizer {
            l2_regularizer {
              weight: 0.0
            }
          }
          initializer {
            variance_scaling_initializer {
              factor: 1.0
              uniform: true
              mode: FAN_AVG
            }
          }
        }
        use_dropout: false
        dropout_keep_probability: 1.0
        conv_hyperparams {
          op: CONV
          regularizer {
            l2_regularizer {
              weight: 0.0
            }
          }
          initializer {
            truncated_normal_initializer {
              stddev: 0.01
            }
          }
        }
        predict_instance_masks: true
        mask_prediction_conv_depth: 0
        mask_height: 33
        mask_width: 33
        mask_prediction_num_conv_layers: 4
      }
    }
    second_stage_post_processing {
      batch_non_max_suppression {
        score_threshold: 0.0
        iou_threshold: 0.6
        max_detections_per_class: 100
        max_total_detections: 100
      }
      score_converter: SOFTMAX
    }
    second_stage_localization_loss_weight: 2.0
    second_stage_classification_loss_weight: 1.0
    second_stage_mask_prediction_loss_weight: 4.0
    resize_masks: false
  }
}
train_config {
  batch_size: 1
  data_augmentation_options {
    random_horizontal_flip {
    }
  }
  optimizer {
    momentum_optimizer {
      learning_rate {
        cosine_decay_learning_rate {
          learning_rate_base: 0.008
          total_steps: 200000
          warmup_learning_rate: 0.0
          warmup_steps: 5000
        }
      }
      momentum_optimizer_value: 0.9
    }
    use_moving_average: false
  }
  gradient_clipping_by_norm: 10.0
  fine_tune_checkpoint: "workspace/pre-trained-models/mask_rcnn_inception_resnet_v2_1024x1024_coco17_gpu-8/checkpoint/ckpt-0"
  num_steps: 2000
  fine_tune_checkpoint_type: "detection"
  fine_tune_checkpoint_version: V2
}
train_input_reader {
  label_map_path: "workspace/annotations/Micro/label.pbtxt"
  load_instance_masks: true
  tf_record_input_reader {
    input_path: "workspace/annotations/Micro/train.record"
  }
  mask_type: PNG_MASKS
}
eval_config {
  metrics_set: "coco_detection_metrics"
  metrics_set: "coco_mask_metrics"
  use_moving_averages: false
  eval_instance_masks: true
  include_metrics_per_category: true
  batch_size: 1
}
eval_input_reader {
  label_map_path: "workspace/annotations/Micro/label.pbtxt"
  shuffle: true
  num_epochs: 1
  load_instance_masks: true
  tf_record_input_reader {
    input_path: "workspace/annotations/Micro/test.record"
  }
  mask_type: PNG_MASKS
}
```

**Comment by TannerGilbert** [comment]
This certainly is strange. I think this might have happend because I specified `batch_size = 1` when originally creating the notebook. I just increased it to 4 inside the notebook, and it seems to work. Hopefully, this will fixe the issue for you too.

**Comment by NyanSwanAung** [comment]
@TannerGilbert  I did change to batch size 4 and weirdly, it took longer to train and the loss is the same as batch size 1

**Comment by NyanSwanAung** [comment]
@TannerGilbert  I've opened another [issue](https://github.com/TannerGilbert/Tensorflow-Object-Detection-API-train-custom-Mask-R-CNN-model/issues/14#issue-914380975). Please check.

**Comment by emitoooo** [comment]
Try to use a lower learning_rate, that should help.
